Add nullability annotations to some JDK APIs.

cpovirk opened the pull request:
Merge together with https://github.com/typetools/jdk/pull/37

- `*Reference`: All constructors accept `null`, and all `get` methods can return `null`. (The `Reference` superclass itself already had an annotation on `get`, but the subclasses did not.)
- `ParameterizedType`: `getOwnerType` can return `null`.
- `URL`: The `host` parameters to the constructors can be `null`.
- `HashMap` and `Map`: `putIfAbsent` and `replace` can return `null`.
- `Objects`: `requireNonNull` ensures that its first argument is non-null, and it returns `@NonNull`, even if `<T>` might allow `null`.
- `PriorityQueue`, `SortedMap`, `SortedSet`, `TreeMap`, `TreeSet`, `ConcurrentSkipListSet`: `comparator` can returns `null`, and it is `@Pure`. Additionally, the `TreeMap` and `TreeSet` constructors accept a null `Comparator`. (`ConcurrentSkipListSet` already did.)
- `Spliterator` (and primitive specializations): `trySplit` and `getComparator` can return `null`, and `getComparator` is `@Pure`.

cpovirk commented:
I'm not sure if any of these will be controversial, so I've held off on creating a corresponding pull request for typetools/jdk. Once this one is merged, I can make the same changes there.

wmdietl commented on checker/jdk/nullness/src/java/util/Map.java:
```diff
@@ -818,7 +818,7 @@ default V putIfAbsent(K key, V value) {
      * @since 1.8
      */
-    default boolean remove(Object key, Object value) {
+    default boolean remove(@Nullable Object key, @Nullable Object value) {
```
Similar `remove` methods, e.g. in `List` https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/util/List.java#L278 are also annotated as `@Nullable Object`.
While a sound solution for this would be preferable, this isn't particular to this PR and can be solved separately.
I would suggest opening an issue to keep track of this soundness issue.

cpovirk replied:
These methods are particularly interesting because it's possible to have, say, a `HashMap<@Nullable String, @Nullable String>`. If the `remove` methods don't accept `null`, then there's no direct way to remove the null values that the API allows users to put in :) (Well, I suppose that they could retain a reference of static type `HashMap`, assuming that `HashMap.remove` is annotated to accept null values.)

As @wmdietl hints, even methods like [`Collection.contains`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#contains(java.lang.Object)) and [`Map.get`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#get(java.lang.Object)) are allowed to reject `null`, which would prevent users from even _reading_ the values they put in.

All that said, you're certainly right that it's unsound. I had figured it was a bridge you've already crossed, but this may deserve more thought before going any further.

Perhaps what we "really" want is:

```java
boolean remove(@NullnessMatchesNullnessOf("K") key, @NullnessMatchesNullnessOf("V") value);
```

wmdietl replied:
I found the previous discussion of this: #1326

cpovirk replied:
Thanks for the link!

If you feel that you're in a hole here and you want to stop digging, I'm happy to revert. (I could even remove the existing `@Nullable` annotations if you'd like, though maybe that's more disruptive to existing users than you'd prefer?)

mernst replied:
How about putting the null-friendly annotations in a separate stub file?  See https://github.com/typetools/checker-framework/issues/3040.

mernst replied:
(That would require moving those changes to a different pull request.)

cpovirk replied:
Will do.

cpovirk commented:
Sorry about the formatting change. I probably assumed that I introduced the stray line when adding imports. Thanks for reverting it. For my reference, I think `hub pr checkout 3029` has pulled your changes back into my client.

I've reverted `ResourceBundle`, which I doubt is frequently used enough to justify going through to annotate the whole thing.

And yeah, I should have been a little more hygienic, rather than using `master`... :)

I'll move on to the inline comments next. As a meta point, I am happy to talk your ear off about opinions about these things, but ultimately we don't overly mind keeping around stub files for cases in which we're willing to play a little more fast and loose with type-safety than the Checker Framework wants to by default. Obviously it is your project, and you can do whatever you'd like with pull requests no matter what, but please do feel especially free to continue pushing back, and we will keep whatever changes we need in our stub files.

cpovirk commented on checker/jdk/nullness/src/java/util/Objects.java:
```diff
@@ -252,5 +254,6 @@
      * @since 1.8
      */
     // TODO: treat like other nullness assertion methods in the Checker Framework.
```
Done (in changes to be pushed when after I've replied to the other comments).

cpovirk commented on checker/jdk/nullness/src/java/util/Objects.java:
```diff
@@ -177,6 +177,7 @@
      * @throws NullPointerException if {@code obj} is {@code null}
      */
+    @EnsuresNonNull("#1")
     @SideEffectFree public static <T> T requireNonNull(@Nullable T obj) { throw new RuntimeException(); }
```
It might as well. Done (again, in changes to be published shortly).

That said: Most of the time, the Checker Framework can probably infer `<T>` to be the appropriate non-null type, in which case the return value of plain `T` would be sufficient. Arguably we might even want to change the type parameter declaration to `<T extends @NonNull Object>` to ensure that this happens. But that might break any existing code that writes `Objects.<X>requireNonNull(x)`, where `<X>` has a nullable bound. (The code would have to change to `Objects.<@NonNull X>requireNonNull(x)`.) That seems fairly unlikely to come up, but maybe it's most consistent to accommodate such code anyway, similar to what you do with the type parameter bound for `Class` and `Optional`.

mernst replied:
> we don't overly mind keeping around stub files for cases in which we're willing to play a little more fast and loose with type-safety than the Checker Framework wants to by default.

I understand.  As much as possible, I would like to reduce your use of those stub files (or migrate some into the Checker Framework proper, as I proposed for the nullFriendlyCollections.astub).  One reason is to reduce your maintenance burden.  Another reason is to give other users the benefit of the annotation work you have done.  The conversations about the annotation tradeoffs are very useful, and I appreciate them.

mernst replied:
Hmm, I'm not sure how that comment ended up in this thread.  What I meant to say is that I'm inclined to keep the lenient bounds.  I'm often surprised by the imaginative ways that programmers use Java. :-)

cpovirk replied:
Coincidentally, I just came across a ton of calls to `Preconditions.<@Nullable Foo>checkNotNull(foo)` in our codebase. I assume they were a holdover from before we had a Checker Framework stub for the class, but still, that's evidence that tightening the bounds would force some users to make changes.

cpovirk commented on checker/jdk/nullness/src/java/lang/ref/SoftReference.java:
```diff
@@ -82,7 +83,7 @@
      * @param referent object the new soft reference will refer to
      */
-    public SoftReference(T referent) {
+    public SoftReference(@Nullable T referent) {
```
I expected to have no real opinion here, but after looking into it a little, I'll advocate further for `@Nullable`.

I don't personally need the constructor to accept `null`. I was just annotating Guava's subclasses ([example](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/FinalizableSoftReference.html)), and I noticed that at least our current version of the JDK supports `null`, so I put the annotation on the Guava and JDK classes.

- I am shocked to discover that ~7% of all `Reference` construction sites in the wider Google codebase pass a literal `null` argument. These mostly appear to be cases in which there will eventually be a "real" `Reference` but the code wants to eagerly initialize a field with _some_ `Reference` object to avoid leaving the field itself `null`. Grepping through the JDK sources shows some instances of the pattern there, too.

- Possibly some other number of `Reference` objects in the Google codebase are constructed with a non-constant value that _might_ be `null`. I doubt this is very common, but it's harder to easily search for. Of course, such users could always perform an explicit null check -- assuming that they don't need to end up with a `Reference` instance to initialize a field or something, as in the previous bullet.

- Here is some additional evidence that the JDK owners at least once upon a time considered null referents to be valid: https://bugs.openjdk.java.net/browse/JDK-4247236 As you note, though, there is no documentation of this that I can see.

Setting that aside for a moment: If you do end up wanting to forbid `null`, then we should probably change the type parameter to `<T extends [@NonNull Object>` so that users can't create a `SoftReference<@Nullable Object>` and pass in a null referent that way.

cpovirk replied:
> If you do end up wanting to forbid `null`, then we should probably change the type parameter to `<T extends [@NonNull Object>` so that users can't create a `SoftReference<@Nullable Object>` and pass in a null referent that way.

(Or we can just annotate the constructor parameter as `@NonNull`. I always forget about that :))

mernst replied:
OK, we can permit `null` as an argument to reference constructors.  I would personally prefer
 * use null instead of a nonsense reference, or
 * create a special nonsense reference, suppress the warning once at its construction, and use it where needed

However, I will bow to actual usage, including in the JDK.
Thanks for providing this extra evidence!

cpovirk commented:
And I took a pass through the files to see if I'd annotated some only partially. I noticed that I need to annotated the public nested classes in `Spliterator`, so I've done that. The rest looked good to me, at least.

cpovirk commented:
I updated the initial post in this thread (and noticed that I'd previously forgotten to mention that I'd handled `PriorityQueue.comparator()`, too). Maybe that will automatically propagate to the commit when this PR is accepted? Regardless, it should be a summary of where this PR stands.

cpovirk commented:
Oops, sorry again: I think I reverted some changes to `Map` that should have stayed. I'll fix that.

cpovirk commented:
OK, I re-added the `@Nullable` annotations on the _return values_ of `Map.putIfAbsent` and `replace`. I also added them on the overrides on `HashMap`. And I updated the first post in this thread.

Sorry for all my mixups.

mernst commented on checker/jdk/nullness/src/java/net/URL.java:
```diff
@@ -307,7 +307,7 @@
      *                  java.lang.String)
      */
-    public URL(String protocol, String host, int port, String file)
+    public URL(String protocol, @Nullable String host, int port, String file)
```
Is this string null for a so-called "relative" URL?  Why aren't all the other formal parameter types also `@Nullable`?  I presume they can also be missing for a relative URL.

cpovirk replied:
I'm not really sure what Java is going for here. As you say, a relative URL should allow omitting other components, like protocol. My impression from the Javadoc is that an instance of `URL` is intended to represent only absolute URLs (though it does support _resolving_ a relative URL into an absolute URL if you pass the relative URL as a `String spec` to a constructor that accepts a `URL context`).

The only official doc I've found for a null host is in [the Javadoc for `equals`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URL.html#equals(java.lang.Object)):

> Two hosts are considered equivalent if both host names can be resolved into the same IP addresses; else if either host name can't be resolved, the host names must be equal without regard to case; or _both host names equal to null_.

As for why I even made this change:

It looks like Guava passes `null` to this constructor to create `file:` URLs [in `ClassPath`](https://github.com/google/guava/blob/c6f48dced81bc0dbfb8df3cfcc9475d6b9e0f460/guava/src/com/google/common/reflect/ClassPath.java#L483). (We've also copied that code to a few locations in our tests.)

I can check whether an empty-string host would work just as well. It might, given that the _parsing_ constructor, `new URL(String)`, turns an input of `file:///home/cpovirk` into an object with a empty-string host, not a null host.

cpovirk replied:
It does look like we can pass an empty string just as well as `null`. That suggests that accepting `null` here isn't essential. Let me know if you'd like me to revert this part.

mernst replied:
This seems OK to me either way; it's not worth delaying the pull request over, and null is a better value to use than the empty string in any event.  Thank you for investigating!

cpovirk commented:
And sorry, I forgot to reply about https://github.com/typetools/jdk/ . I can take care of it, assuming that you don't have some desire to do it yourself :)

cpovirk commented:
Created https://github.com/typetools/jdk/pull/37.

Also, I forgot: Should I remove these comments?

https://github.com/typetools/checker-framework/blob/6dfaf925f01f43def07c0c1f67fcc5333bc3bd2c/checker/jdk/nullness/src/java/util/Objects.java#L254

...and...

https://github.com/typetools/jdk/blob/1d0ed024e808fe7e71d5e3d0280443efb4d8cd07/src/java.base/share/classes/java/util/Objects.java#L237

cpovirk commented:
Answering my own question: [Yes](https://github.com/typetools/checker-framework/pull/3029#discussion_r363135916), I was supposed to remove them. I think that I did at one point but that the deletion got lost in merging [another change to `Objects`](https://github.com/typetools/checker-framework/commit/335b3f7f74d96bc53738a6e36a310f2a1511c8cb#diff-f16013460cdca51c59b0cb05e7697a9b). I've updated this PR and will update the corresponding `jdk` one shortly.

mernst commented:
@cpovirk
When two related pull requests are made, they must be made from corresponding branches of corresponding forks of the `checker-framework` and `jdk` repositories.
However, https://github.com/typetools/checker-framework/pull/3029 is from branch `master` and 
https://github.com/typetools/jdk/pull/37 is from branch `jdk-version-of-3029`.  (Furthermore, either the `jdk` pull request must be made first, or the CI job for the `checker-framework` pull request must be re-triggered after the `jdk` pull request is made.)

Would it be possible for you to re-open the pull requests, using like-named branches?

Thanks!

cpovirk commented:
Sure. It sounds like the shortest route from here to there is to close this PR and submit an identical one using the now mildly unfortunate branch name `jdk-version-of-3029` to match the `jdk` repo. I've done that in https://github.com/typetools/checker-framework/pull/3191. Let me know if I need to do anything differently.
